### PR TITLE
Migrate IAM to aws-sdk-go-v2

### PR DESCRIPTION
**Pull request by rifelpet**
ref: #16424

### CODE REVIEW

File: pkg/model/awsmodel/iam.go [hunk 1]
```diff
@@ -21,8 +21,8 @@ import (
 	"sort"
 	"strings"
 
+	awsIam "github.com/aws/aws-sdk-go-v2/service/iam"
```
Comment (justinsb) [inline pkg/model/awsmodel/iam.go:24]: Nit: we should probably switch `awsIam` -> `awsiam` at some stage, it's more go idiomatic I believe

File: pkg/model/awsmodel/iam.go [hunk 2]
```diff
@@ -473,20 +473,25 @@ func (b *IAMModelBuilder) buildAWSIAMRolePolicy(role iam.Subject) (fi.Resource,
...
+		}
+		for _, role := range page.Roles {
 			if !strings.HasSuffix(fi.ValueOf(role.RoleName), "."+b.Cluster.ObjectMeta.Name) {
 				continue
 			}
 			getRequest := &awsIam.GetRoleInput{RoleName: role.RoleName}
-			roleOutput, err := iamapi.GetRole(getRequest)
+			roleOutput, err := iamapi.GetRole(ctx, getRequest)
 			if err != nil {
-				getRoleErr = fmt.Errorf("calling IAM GetRole on %s: %w", fi.ValueOf(role.RoleName), err)
```
Comment (justinsb) [inline pkg/model/awsmodel/iam.go:488]: Ah, this is going to be a much more ergonomic API :-)

File: pkg/model/awsmodel/iam.go [hunk 3]
```diff
@@ -500,13 +505,9 @@ func (b *IAMModelBuilder) FindDeletions(context *fi.CloudupModelBuilderContext,
 				}
 			}
 		}
-		return true
-	})
+	}
 	if getRoleErr != nil {
```
Comment (justinsb) [inline pkg/model/awsmodel/iam.go:509]: Nit: this isn't used any more, I believe?

File: pkg/resources/aws/aws.go [hunk 4]
```diff
@@ -2078,11 +2078,17 @@ func ListIAMOIDCProviders(cloud fi.Cloud, vpcID, clusterName string) ([]*resourc
...
-				continue
-			} else if err != nil {
-				return nil, fmt.Errorf("error getting IAM OIDC Provider: %v", err)
+			resp, err := c.IAM().GetOpenIDConnectProvider(ctx, descReq)
+			if err != nil {
+				var nse *iamtypes.NoSuchEntityException
+				if errors.As(err, &nse) {
+					klog.Warningf("could not find IAM OIDC Provider %q. Resource may already have been deleted: %v", aws.StringValue(arn), nse)
+					continue
+				} else if awserror, ok := err.(smithy.APIError); ok && awserror.ErrorCode() == "403" {
```
Comment (justinsb) [inline pkg/resources/aws/aws.go:2087]: This feels like a more fragile API than we had before.  Maybe we should hide `if awserror, ok := err.(smithy.APIError); ok && awserror.ErrorCode() == "403" ` behind a function like `IsHTTP403Error(err error) bool` ?

File: upup/pkg/fi/cloudup/awstasks/iamrolepolicy.go [hunk 5]
```diff
@@ -95,27 +100,26 @@ func (e *IAMRolePolicy) Find(c *fi.CloudupContext) (*IAMRolePolicy, error) {
 		PolicyName: e.Name,
 	}
 
-	response, err := cloud.IAM().GetRolePolicy(request)
-	if awsErr, ok := err.(awserr.Error); ok {
-		if awsErr.Code() == iam.ErrCodeNoSuchEntityException {
+	response, err := cloud.IAM().GetRolePolicy(ctx, request)
+	if err != nil {
+		var nse *iamtypes.NoSuchEntityException
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/iamrolepolicy.go:105]: I guess we could also have a helper for this check, but this one is simpler.

File: upup/pkg/fi/cloudup/awsup/aws_cloud.go [hunk 6]
```diff
@@ -333,16 +332,7 @@ func NewAWSCloud(region string, tags map[string]string) (AWSCloud, error) {
...
-			Config:            *config,
-			SharedConfigState: session.SharedConfigEnable,
-		})
-		if err != nil {
-			return c, err
-		}
-		c.iam = iam.New(sess, config)
-		c.iam.Handlers.Send.PushFront(requestLogger)
-		c.addHandlers(region, &c.iam.Handlers)
+		c.iam = iam.NewFromConfig(cfgV2)
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awsup/aws_cloud.go:335]: So we are losing 3 things:
* SharedConfigState 
* requestlogger
* addHandlers - which configures the cross-request delay

Do we think any of those are still important?

**Comment by justinsb** [comment]
Thanks for doing this @rifelpet 

Generally looks good, but my one "blocker" is whether we've lost things by "just" doing `iam.NewFromConfig(cfgV2)`.  I'm fine if the answer is "we may be losing things, and if we are we can add them back, but we don't think we need them" - but I just want to make sure we're doing it consciously.

**Comment by rifelpet** [comment]
> Thanks for doing this @rifelpet
> 
> Generally looks good, but my one "blocker" is whether we've lost things by "just" doing `iam.NewFromConfig(cfgV2)`. I'm fine if the answer is "we may be losing things, and if we are we can add them back, but we don't think we need them" - but I just want to make sure we're doing it consciously.

I mention the config changes [here](https://github.com/kubernetes/kops/pull/16432#issue-2215225398). I'm hoping that the SDK's new built-in retryer and retry logger will be sufficient for our needs but we'll find out once we run scale tests (and likely wont find out until the largest AWS service is migrated - EC2)

I'll address your nits in a followup

File: upup/pkg/fi/cloudup/awsup/aws_cloud.go (hunk 6 above)
Comment (rifelpet) [reply upup/pkg/fi/cloudup/awsup/aws_cloud.go:335]: SharedConfigState is now default: https://aws.github.io/aws-sdk-go-v2/docs/migrating/#client-construction

there is no way to disable or enable SharedConfigState anymore, only specify which shared config file to use (default `~/.aws/config`)

**Pull request merged**